src/lib/blockchain_snark/blockchain_snark_state.ml

```diff
@@ -363,10 +363,14 @@ type tag =
   (State_hash.var, Protocol_state.value, Nat.N2.n, Nat.N1.n) Pickles.Tag.t
 
 module type S = sig
-  module Proof :
-    Pickles.Proof_intf
-      with type t = (Nat.N2.n, Nat.N2.n) Pickles.Proof.t
-       and type statement = Protocol_state.Value.t
+  module Proof : sig
+    include
+      Pickles.Proof_intf
+        with type t = (Nat.N2.n, Nat.N2.n) Pickles.Proof.t
+         and type statement = Protocol_state.Value.t
+
+    val verify : (statement * t) list -> bool Async.Deferred.t
```

Comment:
this kind of monkey-patching has been detrimental to code maintenance in the past. Is there a reason this isn't something we can put directly in the `Pickles.Proof_intf`?

Reply:
Yeah I see what you mean..
I am overwriting `verify (...) -> Promise.t` with `verify (...) -> Deferred.t` to keep `Promise.t` out of `blockchain_snark` as much as possible. In `Pickles.Proof_intf`, pickles' proof interface, we need verify to return `Promise.t` because that interface is used in JS.

So the only alternative I see to this is retain the exact `Pickles.Proof_intf` as the interface for `S` here, but that implies having a much larger part of the code be affected by the move to Promise -- blockchain_snark plus other libs like `prover` which depend on it, and would mean either a mix of Promise and Deferred in those libs or an even larger diff which moves unrelated parts to Promise as well

Reply:
Other alternatives
* inline the entire interface (it's fairly small) instead of monkey-patching the one form pickles
* Make `deferred` another abstract type in the pickles interface
Your pick :D I prefer #1 or the existing solution

Reply:
Secret option 3: a promise can be converted to a deferred with
```ocaml
let res = Ivar.create () in
Promise.bind promise ~f:(fun x -> res) |> ignore
Ivar.read res
```
and then pickles can return a `verify` and `verify_promise`.

Reply:
Yeah I like that. I'll make `to_deferred` part of the `Promise` API, so that both JS and OCaml have it, and add `verify` and `verify_promise`. This also neatly solves the comment below

Reply:
Done!